presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,739 @@
+
+        ListenableFuture<List<SerializedPage>> memoryPages = transformAsync(storagePages, input -> {
+            result.addAll(input);
+            long pageCount = input.size();
+            long bytes = getPagesSize(input);
+            if (bytes < maxBytes || input.isEmpty()) {
+                return getPagesFromMemory(token + pageCount, getTracker);
+            }
+            return immediateFuture(ImmutableList.of());
+        }, executor);
+
+        ListenableFuture<BufferResult> resultFuture = transform(memoryPages, input -> {
+            long newSequenceId = token + input.size();
+            result.addAll(input);
+            return new BufferResult(taskInstanceId, token, newSequenceId, false, result.build());
+        }, executor);
+
+        return catchingAsync(resultFuture, Exception.class, Futures::immediateFailedFuture, executor);
+    }
+
+    @Override
+    public synchronized void acknowledge(OutputBuffers.OutputBufferId bufferId, long token)
+    {
+        checkArgument(bufferId.getId() == 0, "Invalid buffer id");
+        checkArgument(token >= 0, "Invalid token");
+
+        // ignore if buffer is destroyed OR pages have been acknowledged already
+        if (state.get() == FINISHED || token < sequenceId.get()) {
+            return;
+        }
+
+        long oldSequenceId = sequenceId.get();
+        int pagesToRemove = toIntExact(token - oldSequenceId);
+        long currentSequenceId = oldSequenceId;
+
+        checkArgument(pagesToRemove <= totalPagesRemaining.get(), "Invalid token");
+
+        // remove the pages from storage
+        currentSequenceId += acknowledgePagesFromStorage(token);
+
+        // remove the pages from memory
+        if (currentSequenceId < token) {
+            acknowledgePagesFromMemory(token, currentSequenceId);
+        }
+
+        verify(sequenceId.compareAndSet(oldSequenceId, oldSequenceId + pagesToRemove));
+    }
+
+    @Override
+    public void abort(OutputBuffers.OutputBufferId bufferId)
+    {
+        checkArgument(bufferId.getId() == outputBufferId.getId(), "Invalid outputBufferId");
+        destroy();
+    }
+
+    @Override
+    public ListenableFuture<?> isFull()
+    {
+        return immediateFuture(null);
+    }
+
+    @Override
+    public void enqueue(Lifespan lifespan, List<SerializedPage> pages)
+    {
+        PendingRead pendingRead;
+        synchronized (this) {
+            if (!state.get().canAddPages()) {
+                return;
+            }
+
+            this.pages.addAll(pages);
+
+            // update output buffer info
+            totalBufferedBytes.addAndGet(getPagesSize(pages));
+            totalBufferedPages.addAndGet(pages.size());
+            totalPagesAdded.addAndGet(pages.size());
+            totalRowsAdded.addAndGet(getPagesRows(pages));
+
+            totalPagesRemaining.addAndGet(pages.size());
+
+            if (getPagesSize(this.pages) >= threshold.toBytes()) {
```

Comment:
`totalBufferedBytes` refers to the total bytes of the pages in memory and in the temp storage. We're calculating the total bytes of the in-memory pages here to see if it exceeds the threshold we need to flush the pages into temp storage.

Reply:
`getInMemoryPagesSize` wouldn't be an appropriate name, since the method doesn't return the total bytes of the in-memory pages specifically. For example, in `processRead`, we pass in a list of pages we've read from TempStorage to determine how many bytes we've read so far from storage.